src/map.cpp

```diff
@@ -5053,105 +5053,21 @@ std::list<item> map::use_charges(const tripoint &origin, const int range,
-            ret.push_back(tmp);
-
-            if (quantity == 0) {
-                return ret;
-            }
-        }
-
-        if (weldpart >= 0) { // we have a weldrig, now to see what to drain
-            itype_id ftype = "null";
-
-            if (type == "welder") {
-                ftype = "battery";
-            } else if (type == "soldering_iron") {
-                ftype = "battery";
-            }
-
-            item tmp(type, 0); //TODO add a sane birthday arg
-            tmp.charges = veh->drain(ftype, quantity);
-            quantity -= tmp.charges;
-            ret.push_back(tmp);
-
-            if (quantity == 0) {
-                return ret;
-            }
-        }
-
-        if (craftpart >= 0) { // we have a craftrig, now to see what to drain
-            itype_id ftype = "null";
+        // try to provide as many charges as possible from any matching vehicle pseudo-tool
+        if( veh->has_part( p, [&type]( const vehicle_part &e ) { return e.info().tools.count( type ); } ) ) {
+            item obj( type );
+            obj.set_flag( "PSEUDO" );
+            obj.ammo_set( obj.ammo_default(), veh->drain( obj.ammo_default(), quantity ) );
+            quantity -= obj.ammo_remaining();
 
-            if (type == "press") {
-                ftype = "battery";
-            } else if (type == "vac_sealer") {
-                ftype = "battery";
-            } else if (type == "dehydrator") {
-                ftype = "battery";
-            }
-
-            item tmp(type, 0); //TODO add a sane birthday arg
-            tmp.charges = veh->drain(ftype, quantity);
-            quantity -= tmp.charges;
-            ret.push_back(tmp);
-
-            if (quantity == 0) {
-                return ret;
-            }
-        }
-
-        if (forgepart >= 0) { // we have a veh_forge, now to see what to drain
-            itype_id ftype = "null";
-
-            if (type == "forge") {
-                ftype = "battery";
-            }
+            ret.push_back( obj );
 
-            item tmp(type, 0); //TODO add a sane birthday arg
-            tmp.charges = veh->drain(ftype, quantity);
-            quantity -= tmp.charges;
-            ret.push_back(tmp);
-
-            if (quantity == 0) {
+            if( quantity == 0 ) {
                 return ret;
             }
         }
 
-        if (chempart >= 0) { // we have a chem_lab, now to see what to drain
-            itype_id ftype = "null";
-
-            if (type == "chemistry_set") {
-                ftype = "battery";
-            } else if (type == "hotplate") {
-                ftype = "battery";
-            }
-
-            item tmp(type, 0); //TODO add a sane birthday arg
-            tmp.charges = veh->drain(ftype, quantity);
-            quantity -= tmp.charges;
-            ret.push_back(tmp);
-
-            if (quantity == 0) {
-                return ret;
-            }
-        }
+        const int cargo = veh->part_with_feature(vpart, "CARGO");
```

Comment:
Depends - if you're copying or indenting, it's line-free and saves future lines. It only adds lines if you're editing lines that would otherwise stay completely unchanged.